cart/src/endpoints/handlers.js

```diff
@@ -0,0 +1,281 @@
+
+    try {
+      // Add the product to the cart
+      const res = await this.model.addProduct(req.params.id, req.payload);
+      this.logger.debug(`\tResult: ${JSON.stringify(res)}`);
+
+      // Return what was added
+      return rep.response({message: "Product added to cart.", data: res}).code(201);
+    }
+
+    // Catch any database errors (e.g. product not found) and return the appropriate response
+    catch (err) {
+      if (err.constraint === "products_in_cart_pid_fkey") {
+        let message = 'Product does not exist.';
+        this.logger.debug(`\t${message}`);
+        return rep.response({message: message}).code(400);
+      }
+      else if (err.constraint === "carts_uid_fkey") {
+        let message = 'User does not exist.';
+        this.logger.debug(`\t${message} -- ${err.detail}`);
+        return rep.response({message: message}).code(400);
+      }
+      else if (err.constraint === "products_in_cart_pkey") {
+        let message = 'Product already present in cart.';
+        this.logger.debug(`\t${message} -- ${err.detail}`);
+        return rep.response({message: message}).code(400);
+      }
+      else {
+        this.logger.error(JSON.stringify(err));
+        throw err;
+      }
+    }
+  }
+
+  /**
+   * Remove a product from a cart
+   * @async
+   * @param {Hapi.request} req - the request object
+   * @param {object} rep - the response toolkit (Hapi.h)
+   */
+  async removeProduct(req, rep) {
+    this.logRequest(req);
+
+    // Check if request contains a body
+    if (!req.payload) {
+      return rep.response({message: "Body cannot be empty."}).code(400);
+    }
+
+    // Check if request body contains the required values
+    const isValid = Handlers.propsPresent(['pid'], req.payload);
+    if (!isValid.valid) {
+      return rep.response({message: `${isValid.missing} not specified.`}).code(400);
+    }
+
+    this.logger.debug(`\tHandler: Removing product ${req.payload} from cart ${req.params.id}`);
+
+    try {
+      // Remove the product from the cart
+      const res = await this.model.removeProduct(req.params.id, req.payload);
+      this.logger.debug(`\tResult: ${JSON.stringify(res)}`);
+
+      // If no rows were removed (i.e. the products wasn't in cart), respond with a 400.
+      if (res === 0) {
+        return rep.response({message: `Product ${req.payload.pid} not in cart ${req.params.id}.`}).code(400);
+      }
+      else {
+        // Otherwise, return  how many rows were removed
+        return rep.response({message: "Product removed from cart.", data: res}).code(200);
+      }
+    }
+    catch (err) {
+      this.logger.error(JSON.stringify(err));
+      throw err;
+    }
+  }
+
+  /**
+   * Change the amount of a product in a cart.
+   * @async
+   * @param {Hapi.request} req - the request object
+   * @param {object} rep - the response toolkit (Hapi.h)
+   */
+  async changeAmount(req, rep) {
+    this.logRequest(req);
+
+    // TODO: Should we deal with product not in stock?
```

Comment:
Yes. I'm not sure if this should even be the responsibility of the cart, or whether that'll be checked some other way, which is why I left it as a todo. If it is the responsibility of the cart, then at this point it would query the inventory management service in some way. This is something that's up for discussion.

Reply:
okay...let's bring it up in our next meeting